### DISCUSSION TITLE
Tweaking downlink data rates

### DISCUSSION THREAD

**Discussion by andrepuschmann** [srsran/srsRAN_Project#89] [Announcements]
Hey folks,

with the most recent public code push we've introduced a few more features and configuration knobs that allow you to achieve higher data rates. They aren't enabled by default (yet) and are easy to miss so I'd like to share them here in case anyone want's to give them a shot.

### Setup details

But let's start with a baseline benchmark first. For all the experiments here I am using my standard development laptop - running Ubuntu on an Intel i7-1260P. No tweaks or special configuration. The only thing I did was to set the CPU governor to performance. Further I am using a Ettus USRP B200 mini attached to the laptop. As user equipment I am using a Motorola Edge 30 Pro. Downlink traffic is generated using iperf running on the same computer as the gNB in the Docker container hosting open5gs. To generate uplink traffic I am using Magic iPerf that I control from the phone. Both DL and UL iperf use UDP as transport protocol.

### Setting the base

Ok, so let's see what we get with the baseline config using a 20 MHz TDD SISO cell. The config can be found [here](https://github.com/srsran/srsRAN_Project/blob/main/configs/gnb_rf_b200_tdd_n78_20mhz.yml). With this config and gains tuned a bit to match my local setup here I get around 38 Mbit/s and 19 Mbit/s in DL and UL respectively. Not too bad. But maybe we can get a bit more even.

```
           -------------DL----------------|------------------UL--------------------
 pci rnti  cqi  mcs  brate   ok  nok  (%) | pusch  mcs  brate   ok  nok  (%)    bsr
   1 4601   15   28    38M 1195    5   0% |  21.9   28    19M  576   24   4%   300k
   1 4601   15   28    38M 1200    0   0% |  22.0   28    19M  573   27   4%   300k
   1 4601   15   28    38M 1200    0   0% |  22.0   28    19M  574   26   4%   300k
   1 4601   15   28    38M 1200    0   0% |  22.2   28    19M  584   16   2%   300k
   1 4601   15   28    38M 1200    0   0% |  22.3   28    19M  585   15   2%   300k
   1 4601   15   28    38M 1198    2   0% |  22.3   28    19M  586   14   2%   300k
   1 4601   15   28    38M 1200    0   0% |  22.3   28    19M  589   11   1%   300k
   1 4601   15   28    38M 1200    0   0% |  22.4   28    19M  593    7   1%   300k
   1 4601   15   28    38M 1200    0   0% |  22.4   28    19M  593    7   1%   300k
```

### Reducing the SSB period
One of those new config options is the periodicity in which the SSB is broadcast. The default is every 10ms. Since the SSB occupies downlink resources that cannot be used for user traffic a higher SSB periodicity results in lower DL rates. [Here](http://howltestuffworks.blogspot.com/2019/10/5g-nr-synchronization-signalpbch-block.html) you can find some more details. For this example let's set the SSB periodicity to 20ms and see what rates we get. This can be done by adding the `ssb_period` param to the `cell_cfg` section in the config file or by just overwriting the default value on the command line.
`./gnb -c config/gnb_rf_b200_tdd_n78_20mhz.yml cell_cfg --ssb_period=20`

With this I could observe slightly higher DL rates - 1 Mbit/s more to be precise.

```
           -------------DL----------------|------------------UL--------------------
 pci rnti  cqi  mcs  brate   ok  nok  (%) | pusch  mcs  brate   ok  nok  (%)    bsr
   1 4601   15   28    39M 1176   24   2% |  26.0   27    46k   14    3  17%    0.0
   1 4601   15   28    39M 1200    0   0% |  26.2   27    37k   11    1   8%    0.0
   1 4601   15   28    39M 1200    0   0% |  25.7   27    45k   13    4  23%    0.0
   1 4601   15   28    39M 1200    0   0% |  25.4   27    45k   12    4  25%    0.0
   1 4601   15   28    39M 1201    0   0% |  25.0   28    45k   12    5  29%    0.0
   1 4601   15   28    39M 1200    0   0% |  25.5   27    39k   11    1   8%    0.0
   1 4601   15   28    39M 1200    0   0% |  26.5   27    42k   13    0   0%    0.0
   1 4601   15   28    39M 1200    0   0% |  25.7   27    50k   14    1   6%    0.0
   1 4601   15   28    39M 1200    2   0% |  26.5   27    45k   13    0   0%    0.0
   1 4601   15   28    39M 1198    3   0% |  26.3   27    50k   14    0   0%    0.0
```

You might have noticed that the article referenced above also mentions 40ms, 80ms and even 160ms as SSB periods. Those currently can't be set because they require CORESET multiplexing pattern 2 and 3 which we currently don't use. Besides that it's probably also good to not increase the periodicity too much in order to keep cell search/selection duration low.

### Enabling QAM256 in the Downlink

One of the new features added recently is the addtion of QAM256 modulation in the downlink. Under very good channel conditions this allows to send 8 Bits in every modulated symbol instead of 6 Bits using QAM64. The feature can be enabled in the config file by adding a new PDSCH subsection to the cell config.
```
   common_scs: 30                                                  # Subcarrier spacing in kHz used for data.
   plmn: "00101"                                                   # PLMN broadcasted by the gNB.
   tac: 7                                                          # Tracking area code (needs to match the core configuration).
   pci: 1                                                          # Physical cell ID.
+  pdsch:
+    mcs_table: qam256
```

With QAM256 enabled we can increase the DL rate significantly - by 35% to be precise. Nice!

```
           -------------DL----------------|------------------UL--------------------
 pci rnti  cqi  mcs  brate   ok  nok  (%) | pusch  mcs  brate   ok  nok  (%)    bsr
   1 4601   15   27    53M 1199    1   0% |  20.0   27    19M  600    0   0%   300k
   1 4601   15   27    53M 1200    0   0% |  20.0   27    19M  600    0   0%   300k
   1 4601   15   27    53M 1198    2   0% |  20.0   27    19M  599    1   0%   300k
   1 4601   15   27    53M 1200    0   0% |  20.0   27    19M  599    1   0%   300k
   1 4601   15   27    53M 1200    0   0% |  20.0   27    19M  600    0   0%   300k
   1 4601   15   27    53M 1200    0   0% |  20.0   27    19M  600    0   0%   300k
   1 4601   15   27    53M 1200    0   0% |  20.0   27    19M  596    4   0%   300k
   1 4601   15   27    53M 1200    0   0% |  20.0   27    19M  600    0   0%   300k
   1 4601   15   27    53M 1200    0   0% |  20.0   27    19M  599    1   0%   300k
   1 4601   15   27    53M 1200    0   0% |  20.0   27    19M  600    0   0%   300k
   1 4601   15   27    53M 1200    0   0% |  20.0   27    19M  600    0   0%   300k
```

We'll add auto-selection of QAM256 based on the UE's capabilities in one of the upcoming releases. 

### Increasing the cell bandwidth

Up until now we've used the default 20 MHz bandwidth. However, the Ettus [B2xx](https://kb.ettus.com/B200/B210/B200mini/B205mini#Device_Overview) series of devices can support up to 56 MHz analog bandwidth per channel. This is enough to run a 50 MHz cell. So let's give this a try and see what rates we can achieve with a 50 MHz cell. For this to work however we need to tune a few more parameters in the config. In particular we need to increase the sample rate to support the wider bandwidth.

```
--- a/configs/gnb_rf_b200_tdd_n78_20mhz.yml
+++ b/configs/gnb_rf_b200_tdd_n78_20mhz.yml
@@ -7,7 +7,7 @@ ru_sdr:
   device_args: type=b200,num_recv_frames=64,num_send_frames=64  # Optionally pass arguments to the selected RF driver.
   clock:                                                        # Specify the clock source used by the RF.
   sync:                                                         # Specify the sync source used by the RF.
-  srate: 23.04                                                  # RF sample rate might need to be adjusted according to selected bandwidth.
+  srate: 61.44                                                  # RF sample rate might need to be adjusted according to selected bandwidth.
   otw_format: sc12
   tx_gain: 50                                                   # Transmit gain of the RF might need to adjusted to the given situation.
   rx_gain: 60                                                   # Receive gain of the RF might need to adjusted to the given situation.
@@ -15,11 +15,13 @@ ru_sdr:
 cell_cfg:
   dl_arfcn: 632628                                                # ARFCN of the downlink carrier (center frequency).
   band: 78                                                        # The NR band.
-  channel_bandwidth_MHz: 20                                       # Bandwith in MHz. Number of PRBs will be automatically derived.
+  channel_bandwidth_MHz: 50                                       # Bandwith in MHz. Number of PRBs will be automatically derived.
   common_scs: 30                                                  # Subcarrier spacing in kHz used for data.
   plmn: "00101"                                                   # PLMN broadcasted by the gNB.
   tac: 7                                                          # Tracking area code (needs to match the core configuration).
   pci: 1                                                          # Physical cell ID.
+  pdsch:
+    mcs_table: qam256
```

Running the gNB with this we can get a whopping 112 Mbit/s in the downlink. Not too bad for a tiny B200mini on a standard laptop.

```
           -------------DL----------------|------------------UL--------------------
 pci rnti  cqi  mcs  brate   ok  nok  (%) | pusch  mcs  brate   ok  nok  (%)    bsr
   1 4602   15   27   111M 1158   42   3% |  21.6   26    74k   12    6  33%    0.0
   1 4602   15   27   112M 1200    1   0% |  20.3   26    51k   16    2  11%    0.0
   1 4602   15   27   111M 1156   44   3% |  20.0   26    40k   11    2  15%    0.0
   1 4602   15   27   112M 1200    0   0% |  21.4   26    43k   15    3  16%    0.0
   1 4602   15   27   112M 1201    0   0% |  20.4   27    79k   20    3  13%    0.0
   1 4602   15   27   112M 1200    1   0% |  20.4   25    46k   13    8  38%    0.0
   1 4602   15   27   112M 1201    0   0% |  22.0   25    52k   24   16  40%    0.0
   1 4602   15   27   112M 1200    0   0% |  19.7   25    40k   13   20  60%    0.0
```

Here is a screenshot of NSG showing the statistics as seen from the phone.

<img src="https://github.com/srsran/srsRAN_Project/assets/525775/0482fd32-f894-4061-a79d-6877835c0930" width="250">

### Next steps

For this configuration with a single channel SDR in TDD mode there aren't any further big increases to be expected. With a different TDD configuration we could certainly amplify the downlink a bit further.

Please give this a try and let us know how it works for you.

Happy hacking!

**Comment by andrepuschmann** [comment]
A quick note on the tuning tips for FDD: the above mentioned tips are of course also applicable for FDD configurations. But be aware that many phones have limited bandwidth support for FDD. The Edge 30 Pro used for example seems to only support up to 40 MHz bandwidth in FDD. And that also only in band n1 and n7. Band n8 and n20 are limited to 20 MHz. With [this](https://gist.github.com/andrepuschmann/cf38e68e59ce9528cc2a20457d1e40f5) config the achievable rate goes up to almost 190 Mbit/s.

```
           -------------DL----------------|------------------UL--------------------
 pci rnti  cqi  mcs  brate   ok  nok  (%) | pusch  mcs  brate   ok  nok  (%)    bsr
   1 4601   14   27   172M  911   84   8% |  11.2   17    32M  427   58  11%   150k
   1 4601   14   27   157M  841  152  15% |  11.1   18    21M  285   86  23%    0.0
   1 4601   15   27   181M  962   44   4% |  10.5   18    17k    4   15  78%    0.0
   1 4601   15   27   188M  995    5   0% |   6.0   11    21k    5    0   0%    0.0
   1 4601   15   27   188M  995    6   0% |   5.8   11    21k    5    0   0%    0.0
   1 4601   15   27   188M  995    6   0% |   5.8   11    21k    5    0   0%    0.0
   1 4601   15   27   188M  995    5   0% |   5.9   11    21k    5    0   0%    0.0
```

<img src="https://github.com/srsran/srsRAN_Project/assets/525775/1c991b8f-d0e2-491d-b4fc-813fcd994417" width="250">

**Reply by andrepuschmann** [reply]
I've updated the config file to be compliant with the newest release as some config params have changed places. You'll find it [here](https://gist.github.com/andrepuschmann/2f242af8578b1cafea4acdd24597396a). Now it's doing 200+Mbit/s

<img src="https://github.com/srsran/srsRAN_Project/assets/525775/00ebf2bf-3678-4c2f-98cb-339af40ea5f9" width="250">

**Comment by gustavobsch** [comment]
Great, thanks for this!

I have a Ettus USRP B200  and when I tested the last config attached, where you reach 200mbit, I'm getting buffer underflows and the UE never attaches..
Reducing Reducing the SSB and enabling QAM256 did increase the speed
Thanks again for the enhancements

**Reply by ismagom** [reply]
Hi @gustavobsch, can you try with this config?

```
ru_sdr:
  device_driver: uhd                                            # The RF driver name.
  device_args: type=b200,num_recv_frames=64,num_send_frames=64  # Optionally pass arguments to the selected RF driver.
  clock: external                                                       # Specify the clock source used by the RF.
  sync:                                                         # Specify the sync source used by the RF.
  srate: 46.08                                                  # RF sample rate might need to be adjusted according to selected bandwidth.
  otw_format: sc12
  tx_gain: 79                                                   # Transmit gain of the RF might need to adjusted to the given situation.
  rx_gain: 60
```

**Reply by andrepuschmann** [reply]
I've updated the full config with those args [here](https://gist.github.com/andrepuschmann/ecbb807abb948f99bd9b5cde5035489b)

**Reply by wmwwimor** [reply]
How about zeromq ? How to increase the rate?
